### ISSUE TITLE
思考如何用Control Gate实现不同层面下的动态稀疏性和对于稀疏性的predictor实现

### ISSUE THREAD

**Issue by lidongyue12138** [acada-sjtu/EdgeTraining#16]
- 研究Boosting那篇predictor怎么训练的，复现
- peng那篇论文运用到剪枝上，Control Gate变成0/1
- 怎么把Control knob**加进入就端到端训练**
  - 不同level的剪枝区别，knob的区别

**Comment by rogercmq** [comment]
1. 源码中的predictor是怎么训练的，对应的label是什么
2. 运用到剪枝上，Control knob变成0/1，binary scalar怎么训练
3. 怎样端到端训练 Control knob，在哪个epoch开始和网络参数一起训练

**Comment by ACA-Lab-SJTU** [comment]
梦奇，调整一下你的目标：
- [ ] 掌握Boosting文章源码中的predictor是怎么训练的，对应的label是什么等，在Github上分享结果并进一步讨论；
- [ ] 根据讨论结果，决定是否尝试复现源算法，观察training/test error, mask程度（被mask的channel数量），实际计算量（数据量）的趋势；
- [ ] 寻找一种针对kennel/channel/filter普世的Control knob机制（变成0/1，或者scaler+阈值），解决其训练问题，是否能达到更好的效果；

*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by ACA-Lab-SJTU** [comment]
东岳，调整一下你的目标：
- [ ] 掌握Boosting文章源码中的predictor是怎么训练的，对应的label是什么等，在Github上分享结果并进一步讨论；
- [ ] 根据讨论结果，决定是否尝试复现源算法，观察training/test error, mask程度（被mask的channel数量），实际计算量（数据量）的趋势；
- [ ] 参考AXNet中用神经网络控制另一个神经网络的思路，试图寻找一个预测稀疏程度和结构的predictor，通过控制knob在训练过程中动态Mask网络；收敛后得到一个经过裁剪的网络

*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by lidongyue12138** [comment]
Boosting中Predictor的训练方法：
1. 将原网络中的卷积层替换成Boosting文章的网络结构，而且卷积层就保留原预训练网络的参数（Our method begins by first replacing all convolutional layer computations with (5), and initializing the new convolutional kernels with previous parameters. ）
2. 在此基础上，将剪枝率d设置成为1（相当于不作任何修剪），微调现有的模型（Initially, we do not suppress any channel computations by using density d = 1 in (6) and fine-tune the resulting network.）
3. 渐渐改变d，例如每次减少10%，这样地迭代的训练，直到达到预期的加速比（we then follow Liu et al. (2017) by iteratively decrementing the overall density d of the network by 10% in each step, and thus gradually using fewer channels to sweep the accuracy/performance trade-off.）

**Comment by rogercmq** [comment]
一张image经过一个卷积核得到一个channel，在每个channel后面添加一个binary的control knob是不能用梯度训练的，我还是想考虑scalar+阈值的方法。

ECCV2018会议里Han Song的paper提出了一种强化学习决策剪枝的state-of-the-art方法，晨宇可以看一下。值得一提的是之前介绍的boosting文章在top-5 error rate上比Han Song的state-of-the-art低了0.1个百分点。

**Comment by ACA-Lab-SJTU** [comment]
@yangcy 看一下

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by lidongyue12138** [comment]
补充：Boosting那篇论文在训练的时候在目标函数中加入了每一层的Lasso正则项

**Comment by lidongyue12138** [comment]
上周扫了一些paper，是和Boosting以及Peng学长论文有点类似的，准备仔细看一下打标的paper：
论文列表

Dynamic Execution 

【ICLR 2019】Dynamic Channel Pruning: Feature Boosting and Suppression 

Spatial Sparsity in CNN 

- 【CVPR 2017】More is Less: A more complicated network with less inference complexity ⭐️⭐️
- 【CVPR 2017】Spatially adaptive computation time for residual networks ⭐️
- 【ICML 2016】Dynamic capacity networks ⭐️⭐️⭐️
- 【CVPR 2018】SBNet: Sparse blocks network for fast inference (sparse activation) ⭐️ 
- 【CVPR 2018】BlockDrop: Dynamic inference paths in residual networks. (RL)
- 【2018】Dynamic deep neural networks: Optimizing accuracy-efficiency trade-offs by selective execution. (RL) ⭐️⭐️
- 【2017】Changing model behavior at test-time using reinforcement learning
- 【NIPS 2017】Runtime neural pruning ⭐️
- 【2017】Outrageously large neural networks: The sparsely-gated mixture-of-experts layer⭐️⭐️⭐️
- 【ICML 2017】Adaptive neural networks for efficient inference ⭐️⭐️
- 【CoRR 2018】Channel gating neural networks ⭐️

Comparison

- 【ICCV 2017】Learning efficient convolutional networks through network slimming
- 【IJCAI 2018】Soft filter pruning for accelerating deep convolutional neural networks 
- 【NIPS 2018】Discrimination-aware channel pruning for deep neural networks 
- 【CVPR 2017】More is less: A more complicated network with less inference complexity 
- 【CoRR 2018】Channel gating neural networks 
- 【ICCV 2017】Channel pruning for accelerating very deep neural networks 
- 【NIPS 2016】PerforatedCNNs: Acceleration through elimination of redundant convolutions 
- 【NIPS 2017】Runtime neural pruning 
- 【ICCV 2017】Channel pruning for accelerating very deep neural networks 
- 【ECCV 2018】AMC: AutoML for model compression and acceleration on mobile devices 
- 【ICCV 2017】ThiNet: A filter level pruning method for deep neural network compression 

Prof. Jiang Assigned Papers: 04/03/2019

- Z. He, A. S. Rakin, and D. Fan. Parametric noise injection: Trainable randomness to improve deep neural network robustness against adversarial attack. CVPR, 2019 (Accepted)
  - trainable Gaussian noise injection at each layer on either activation or weights through solving the min-max optimization problem, embedded with adversarial training
  - PNI method improves both clean- and perturbed-data accuracy in comparison to the state-of-the-art defense methods
- Z. He and D. Fan. Simultaneously optimizing weight and quantizer of ternary neural network using truncated gaussian approximation. CVPR, 2019 (Accepted) ⭐️
  - we propose a novel ternarized neural network training method which simultaneously optimizes both weights and quantizer during training
  - incorporate the thresholds of weight ternarization into a closed-form representation using the truncated Gaussian approximation, enabling simultaneous optimization of weights and quantizer through back-propagation training
- In-situ AI: Towards Autonomous and Incremental Deep Learning for IoT Systems (University of Florida, Chongqing University, Capital Normal University)
  - Rather than constantly moving a tremendous amount of raw data to the Cloud, it would be beneficial to leverage the emerging powerful IoT devices to perform the inference task.
  - We equip deep learning based IoT system with autonomous IoT data diagnosis (minimize data movement), and incremental and unsupervised training method (tackle the big raw IoT data generated in ever-changing in-situ environments). 
- A Shift Towards Edge Machine-Learning Processing. (Google) 
  - 未找到文章
- Computation Reuse in DNNs by Exploiting Input Similarity
  - Based on this observation, we propose a technique to reuse some results of the previous execution, instead of computing the entire DNN. Computations related to inputs with negligible changes can be avoided with minor impact on accuracy, saving a large percentage of computations and memory accesses.
- SnaPEA: Predictive Early Activation for Reducing Computation in Deep Convolutional Neural Networks ⭐️⭐️
  - we propose a predictive early activation technique, dubbed SnaPEA. This technique cuts the computation of convolution operations short if it determines that the output will be negative. 
- Prediction based Execution on Deep Neural Networks⭐️⭐️⭐️
  - we propose a two-stage, prediction-based DNN execution model without accuracy loss
- moDNN: Memory Optimal DNN Training on GPUs
  - This paper proposes an moDNN framework to optimize the memory usage in DNN training. 
- C-GOOD: C-code Generation Framework for Optimized On-device Deep Learning. ICCAD, 2018 (SNU) 
  - The framework is facilitated with various options for software optimization that can be performed according to the optimization methodology proposed in this paper. 
  - Another benefit is that it can generate various styles of C code
- Towards Memory Friendly Long-Short Term Memory Networks (LSTMs) on Mobile GPUs
  - In this study, we aim to explore the memory friendly LSTM on mobile GPUs by hierarchically reducing the off-chip memory accesses. 
- GeneSys: Enabling Continuous Learning through Neural Network Evolution in Hardware. (Georgia Tech)⭐️

**Comment by ACA-Lab-SJTU** [comment]
文章很多，我建议大家能维护一张表，把这些个文章的异同点做一个归纳。表项可以包括目的，任务（推理训练），方法，是否重新练，控制机制等等…

光扫论文不动手实践也不行

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by lidongyue12138** [comment]
第八周更新：
- 跑Boosting论文的代码，把Boosting放在直接开始训练阶段，探究不同的稀疏程度，不同阶段网络的表现与计算量的变化
  - 可以去验证：比如用一个70%的网络直接训练是不是就比剪出来的效果不一样
  - 参考Rethinking的文章
- 继续此方向：参考AXNet中用神经网络控制另一个神经网络的思路，试图寻找一个预测稀疏程度和结构的predictor，通过控制knob在训练过程中动态Mask网络；收敛后得到一个经过裁剪的网络
- 保持看论文：根据论文List，扫完论文后注意更新论文总结的表格

**Comment by rogercmq** [comment]
@ACA-Lab-SJTU  报告老师我又更新了一下上周的工作和接下来的计划，简单来说依然在跟进图森科技的那篇论文，具体写成了周报在 [github weekly report](https://github.com/acada-sjtu/EdgeTraining/blob/master/Doc/Weekly-Report/Mengqi%20Cao-week15.md)

**Comment by ACA-Lab-SJTU** [comment]
我看了一下，这些文章比较有水分。而且仍然是一个prune的文章，并不是针对训练加速的。我记得这个在开会讨论时反复确认过。

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by rogercmq** [comment]
这篇文章的实验部分确实难以让人信服，但是故事很简单：不需要预训练模型，直接在网络训练过程中通过channel-wise scalar衡量output channel的重要性，scalar为零说明对应的channel可以被剪枝。老师我现在有个不太明白的地方，如果我只将scalar不为零的对应output channel传到下一层（可能涉及到动态图的部分）是不是也能达到加速训练的目的呢？

**Comment by rogercmq** [comment]
如果把这个scalar做一下01二值化（借鉴神经网络量化的训练方法），就有点像老师您之前assign给我的 **寻找一种针对kennel/channel/filter普世的Control knob机制** 的任务了

**Comment by ACA-Lab-SJTU** [comment]
没错。我到群里分享一下一篇最新的paper

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by ACA-Lab-SJTU** [comment]
之前几个issue都放弃了对吧？


-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University
